Add generic DataSource feature gate

bswartz opened the pull request:
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

Allow any custom resource to be the data source of a PVC.

This is an alpha feature enabled by the GenericPVCDataSource feature gate.

**Which issue(s) this PR fixes**:

Fixes #88635

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
Added GenericPVCDataSource feature gate to enable using arbitrary custom resources as the data source for a PVC.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
[KEP]: https://github.com/kubernetes/enhancements/commit/933a044fa4efc137cd6a6479c96cbf627a4e92cc
```

bswartz commented:
/sig storage
/assign saad-ali

msau42 commented:
/assign @msau42 @j-griffith

msau42 commented:
@kubernetes/sig-storage-api-reviews

msau42 commented:
/assign @xing-yang

msau42 commented on pkg/apis/core/validation/validation.go:
```diff
@@ -1930,16 +1930,18 @@ func ValidatePersistentVolumeClaimSpec(spec *core.PersistentVolumeClaimSpec, fld
-		}
-		if !supportedDataSourceAPIGroupKinds[groupKind] {
-			allErrs = append(allErrs, field.NotSupported(fldPath.Child("dataSource"), groupKind.String(), groupKindList))
+		if !utilfeature.DefaultFeatureGate.Enabled(features.GenericPVCDataSource) {
```
We can't add feature gate checking to validation in order to handle cases where someone persists an object with the new field, but then disables the feature later, or rolls back to a version with the feature disabled. Checking the feature gate in that case would cause updates/deletion to fail because the persisted object now contains invalid data.

Instead, we rely on `DropDisabledFields` to do the right thing before we get to validation.

https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#adding-unstable-features-to-stable-versions

I think the right thing to do here is to to remove the check for supported groupKinds, and only validate syntax.

bswartz replied:
Thanks Michelle I understand what you mean

msau42 commented:
/label api-review

msau42 commented:
/assign @thockin

j-griffith reviewed (changes requested):
I think there are some details missing here, you'll at least need to address:
1. What happens if this new "open it all up" feature is enabled but SnapshotDataSource is not?
2. The godocs in types.go needs to be updated

bswartz commented on pkg/features/kube_features.go:
```diff
@@ -554,6 +554,12 @@ const (
+	// alpha: v1.18
+	//
+	// Enables usage of any object for volume data source in PVCs
+	GenericPVCDataSource featuregate.Feature = "GenericPVCDataSource"
```
I'm open to suggestions on the specific feature gate name.

I know what you mean about this feature gate implicitly enabling the other feature gates, but I'm not sure how that works at a technical level. Currently, if you enable this feature gate, it doesn't matter whether the other gates are enabled or not because the behavior of the new one overrides them (see unit tests that prove this to be true). However I'm not sure what more we can do to make this clear. Should we generate an error if conflicting feature gates are set?

j-griffith replied:
> Currently, if you enable this feature gate, it doesn't matter whether the other gates are enabled or not because the behavior of the new one overrides them

That's exactly what we DO want IMO, I didn't see the unit tests that actually test that path though (eg snapshots work when VolumeSnapshot feature is disabled but Generic is enabled).  Looking again I see how you handled that, it just wasn't explicitly clear the first time I looked at it.

xing-yang replied:
VolumeSnapshotDataSource remains beta in 1.18.  By enabling this generic feature gate, it implies VolumeSnapshotDataSource is always enabled?  I think we still need to check VolumeSnapshotDataSource feature gate even if this new generic feature gate is enabled.

msau42 replied:
Discussed with @thockin. The behavior of the new generic datasource should supercede the other two feature gates. So I think the way it is implemented is fine. Agree with @j-griffith that there could be more test cases explicitly testing this.

xing-yang replied:
Will this affect our E2E tests?

bswartz replied:
@msau42 I did cover those specific cases. I think @j-griffith agrees. Can you describe the combination that's not tested?

j-griffith replied:
@bswartz Actually I was in favor of testing all permutations :)  That being said, you cover the important ones here I think; and hopefully https://github.com/kubernetes/kubernetes/pull/88686 will land and make half of this irrelevant.

That being said:
1.  I still think the naming is horrible and will cause confusion
2.  You need to update the docstrings in the types files
3.  Additional tests are needed, but they can be added later as this is alpha proposal IMO

bswartz replied:
Okay I added more permutations. Is "AnyVolumeDataSource" the feature gate string you prefer? I'm okay with that one if there are no objections.

j-griffith replied:
that's fine with me, I just want to use the same naming pattern / convention and leave PVC out of it (GenericPVCDataSource) is too close to "VolumePVCDataSource" and doesn't make sense against things like "VolumeSnapshotDataSource".  I know, it seems silly but I think it matters.

bswartz commented on pkg/api/persistentvolumeclaim/util_test.go:
```diff
@@ -297,3 +303,104 @@ func TestPVCDataSourceSpecFilter(t *testing.T) {
+			snapshotEnabled: true,
+			want:            nil,
+		},
+	}
```
The idea is that if the new generic data source feature gate is enabled, it doesn't matter what the other 2 feature gates are set to, because anything is allowed. In theory these 3 feature gates could be set 8 different ways, but it's not worth testing all the combinations. I focused on the worthwhile combinations here with my tests.

j-griffith replied:
> The idea is that if the new generic data source feature gate is enabled, it doesn't matter what the other 2 feature gates are set to

Agreed, but I would argue that it is worth testing all of the combinations.

j-griffith reviewed (changes requested):
If others are "ok" with the level of testing that's fine; The doc additions however have to be added.

bswartz commented:
Okay I think I addressed all of @j-griffith's comments. Thanks for the thorough review!

Please let me know if you see any remaining issues.

bswartz commented:
/retest

bswartz commented:
/retest

j-griffith commented on pkg/apis/core/types.go:
```diff
@@ -419,13 +419,18 @@ type PersistentVolumeClaimSpec struct {
-	// * An existing PVC (PersistentVolumeClaim)
-	// In order to use either of these DataSource types, the appropriate feature gate
-	// must be enabled (VolumeSnapshotDataSource, VolumePVCDataSource)
+	// This field can be used to specify that the volume should initially contain
```
maybe something like:

```
	// This field can be used to specify either:
	// * An existing VolumeSnapshot object (snapshot.storage.k8s.io/VolumeSnapshot - Beta)
	// * An existing PVC (PersistentVolumeClaim)
        // * Any external CRD/Object that implements data population (anyVolumeDataSource - Alpha)
	// In order to use VolumeSnapshot object types, the appropriate feature gate
	// must be enabled (VolumeSnapshotDataSource or AnyVolumeDataSource)
	// If the provisioner or an external controller can support the specified data source, it will create
	// a new volume based on the contents of the specified data source.
	// If the specified data source is not supported, the volume will
	// not be created and the failure will be reported as an event.
	
```

j-griffith reviewed (changes requested):
Minor suggesting to fix up the docstring in types.go.  I think you also need to update the copy in staging/src/k8s.io/api/core/v1/types.go

msau42 commented on pkg/apis/core/validation/validation.go:
```diff
@@ -1929,17 +1923,8 @@ func ValidatePersistentVolumeClaimSpec(spec *core.PersistentVolumeClaimSpec, fld
-		if !supportedDataSourceAPIGroupKinds[groupKind] {
-			allErrs = append(allErrs, field.NotSupported(fldPath.Child("dataSource"), groupKind.String(), groupKindList))
+		if len(spec.DataSource.Kind) == 0 {
+			allErrs = append(allErrs, field.Required(fldPath.Child("dataSource", "kind"), ""))
```
Hm should still validate that the group is required for any kind that is not PersistentVolumeClaim?

bswartz replied:
Yes that would be a worthwhile additional check.

msau42 commented on pkg/apis/core/types.go:
```diff
@@ -417,19 +417,15 @@ type PersistentVolumeClaimSpec struct {
+	// * An existing VolumeSnapshot object (snapshot.storage.k8s.io/VolumeSnapshot - Beta)
+	// * An existing PVC (PersistentVolumeClaim)
+	// * An existing custom resource/object that implements data population (Alpha)
+	// In order to use VolumeSnapshot object types, the appropriate feature gate
```
Do we also need to talk about the cloning case?

msau42 commented on pkg/apis/core/types.go:
```diff
@@ -417,19 +417,15 @@ type PersistentVolumeClaimSpec struct {
-	// If the provisioner does not support the specified data source, the volume will
+	// This field can be used to specify either:
+	// * An existing VolumeSnapshot object (snapshot.storage.k8s.io/VolumeSnapshot - Beta)
+	// * An existing PVC (PersistentVolumeClaim)
```
@j-griffith do you plan on updating alpha-beta designation for cloning in your PR?

msau42 replied:
I see cloning is GA now.

msau42 reviewed (commented):
Generally lgtm, just some minor nits on the comment.

Can you rebase and squash?

bswartz commented:
I rebased, no squash yet. Let me do that now.

bswartz committed:
Add AnyVolumeDataSource feature gate

Allow any custom resource to be the data source of a PVC, if the
AnyVolumeDataSource feature gate is enabled. This is an alpha feature.

bswartz commented:
Had to push one more time to remove test cases where PVC cloning feature gate was disabled. It's squashed now.

msau42 commented:
/lgtm

bswartz commented:
/retest

bswartz commented:
@msau42 I had to push another commit to add these 3 files generated by "make update": 

api/openapi-spec/swagger.json
staging/src/k8s.io/api/core/v1/generated.proto
staging/src/k8s.io/api/core/v1/types_swagger_doc_generated.go

msau42 commented:
/lgtm

thockin commented:
/approve

bswartz commented:
/retest
